---
title: "Identification of regulatory links between transcription and RNA processing with long-read sequencing"
subtitle: "STAR Protocols release"
output:
  rmdformats::readthedown:
    toc: 3
    toc_float: true
    highlight: kate
  css: mycss.css
---


```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(dplyr)
require(GenomicFeatures)
require(GenomicRanges)
require(Biostrings)
require(AnnotationDbi)
require(BSgenome.Dmelanogaster.UCSC.dm6)
require(LATER)
require(LASER)
```

## About the Documentation

All tools required for transcriptional couplings can be found in this R packages.  

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("hilgers-lab/LATER")
devtools::install_github("hilgers-lab/LASER")
```

## Preparing the data. 

Data analysis starts from the bam files from Long-read sequencing data. Files can be produced using minimap2. 
```{bash, eval=FALSE}
minimap2 -ax splice -u f genome.fa long_read.fastq.gz | samtools sort -@ 4 -o output.bam
```

## LATER 

### 5′-3′ database creation

```{r, eval=FALSE}
refExons <- "dm6.ref.gtf"
isoformData <- prepareIsoformDatabase(refExons,
             tss.window=50,
             tes.window=150)
```

### Complementing reference annotation with databases

```{r, eval=FALSE}
refTSS <- "TSS_reference_database_dmel.bed"
isoformData <- addPromoterDatabase(refTSS, ref_tss_annot,
reference_annotation,
window = 50)
```

### Counting full length reads 
```{r, eval=FALSE}
bamPath <- system.file("exdata/testBam.bam", package = 'LATER')
countData <- countLinks(bamPath, isoformData)
```

To explore the reads on IGV is possible to subset the alignment `.bam` file using the read ids. Export the read ids using: 
```{r, eval=FALSE}
readr::write_tsv(readAssignments(countData), "read_assignments.txt")
```

Then go to bash terminal and subset the bam file using samtools command:
```{bash, eval=FALSE}
samtools view -N read_assignments.txt -o filtered_output.bam output.bam
```

### Estimate promoter dominance 

Promoter dominance can be estimated using the following code: 

```{r, eval=FALSE}
gene_bias_estimates <- estimatePromoterDominance(countData, isoformData, method="chisq")
```

## Data can be explored using following functions 

```{r, eval=FALSE}
results(gene_bias_estimates)
dominance(gene_bias_estimates)
```


### Additional documentation 
Additional LATER documentation can be access via 
```{r, eval=FALSE}
vignette("LATER")
```

## LASER 
For LASER detailed documentation and explanations go [here](https://hilgers-lab.github.io/LASER/docs/LASER.html)








