---
title: "ciaTaiLoR"
subtitle: "Feature extraction CIA Transcriptome"
output:
  rmdformats::readthedown:
    toc_float: true
    highlight: kate
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(dplyr)
require(GenomicFeatures)
require(GenomicRanges)
require(Biostrings)
require(AnnotationDbi)
require(BSgenome.Dmelanogaster.UCSC.dm6)
require(LATER)
```


```{r, warning=FALSE}
library(ciaTaiLoR)
```

## Data

CIA Assembly is available within ciaTaiLoR: 

```{r, eval=FALSE}
annot_path <- system.file("extdata/CIA.assembly.allTissues59K.gtf.gz", package="ciaTaiLoR")
cia_annot <- rtracklayer::import.gff(annot_path, feature.type = "exon")
# remove novel Genes for simplificity 
cia_annot <- cia_annot[!cia_annot$gene_id =="novelGene"]
```

```{r, echo=FALSE, include=FALSE}
annot_path <- "/data/hilgers/group2/alfonso/ciaTaiLoR/extdata/CIA.assembly.allTissues59K.gtf.gz"
cia_annot <- rtracklayer::import.gff(annot_path, feature.type = "exon")
# remove novel Genes for simplificity 
cia_annot <- cia_annot[!cia_annot$gene_id =="novelGene"]
#' prepare data for 5'-3' window filtering
#' Create a reference database of all TSS-TES link associations based on annotation
#' @param annotation reference annotation GenomicRanges object
#' @param promoter_window window to consider distinct TSS
#' @param pas_window window to consider distinct TES
#'
#' @return promoterDatabase with classification of links
#' @export
#' @import dplyr GenomicFeatures GenomicAlignments S4Vectors
getIsoformFeatures <- function(annotation, promoter_window, pas_window) {
  strandSort <- function(x) {
    c(
      GenomicRanges::sort(x[x@strand == "+"], decreasing = FALSE),
      GenomicRanges::sort(x[x@strand == "-"], decreasing = TRUE)
    )
  }
  # Build 5'-3' links data base
  # Exon ids
  txdb <- GenomicFeatures::makeTxDbFromGRanges(annotation)
  ebt <- GenomicFeatures::exonsBy(txdb, by = "tx", use.names = TRUE)
  t2g <- AnnotationDbi::select(txdb,
                               keys = names(ebt),
                               keytype = "TXNAME",
                               columns = "GENEID")
  e2 <- BiocGenerics::unlist(ebt)
  e2$transcript_id <- names(e2)
  e2$gene_id = t2g$GENEID[match(e2$transcript_id, t2g$TXNAME)]
  e2$exon_id <- e2$exon_name
  e2$exon_name <- NULL
  e2$type <- "exon"
  names(e2) <- NULL
  mcols(e2) <- mcols(e2)[, c("exon_id", "exon_rank",
                             "transcript_id", "gene_id", "type")]
  bins <- list()
  # TSS data base
  # take first position per transcript and make it single nt
  tss.bins  <-
    strandSort(plyranges::mutate(plyranges::anchor_5p(dplyr::filter(e2, exon_rank ==
                                                                      1)), width = 1))
  # make unique TSS starts merging in a 50nt window.
  tss.base <-
    strandSort(
      GenomicRanges::makeGRangesFromDataFrame(
        reshape::melt(GenomicRanges::reduce(
          GenomicRanges::split(tss.bins, ~gene_id),
          min.gapwidth = promoter_window)),
        keep.extra.columns = TRUE
      )
    )
  bins$tss.bins <- tss.bins
  bins$tss.base <- tss.base
  tss.base <-
    tibble::as_tibble(tss.base)  %>%  dplyr::group_by(value.group_name) %>%
    dplyr::mutate(count =  paste0(value.group_name, ":P", sprintf("%02d", sequence(dplyr::n(
    ))))) %>%
    GenomicRanges::makeGRangesFromDataFrame(., keep.extra.columns = TRUE)
  # annotate isoforms with promoter_id
  ii <- GenomicRanges::findOverlaps(tss.bins, tss.base, maxgap = promoter_window - 1)
  tss.bins.annot <-
    GenomicRanges::makeGRangesFromDataFrame(rbind(data.frame(tss.bins[S4Vectors::queryHits(ii)],
                                                             tss.base[S4Vectors::subjectHits(ii)])), keep.extra.columns = TRUE)
  tss.bins.annot <-
    tss.bins.annot %>% dplyr::mutate(transcript_id = transcript_id, promoter_id =
                                       count) %>%
    plyranges::select(gene_id, transcript_id, promoter_id)
  # TES data base
  # last exon per transcript
  le <-
    GenomicRanges::makeGRangesFromDataFrame(
      e2 %>% group_by(transcript_id) %>% dplyr::filter(exon_rank  == max(exon_rank)),
      keep.extra.columns = TRUE
    )
  tes.bins  <-
    strandSort(plyranges::mutate(plyranges::anchor_3p(le), width = 1))
  # make unique TSS starts merging in a 50nt window.
  tes.base <-
    strandSort (
      GenomicRanges::makeGRangesFromDataFrame(
        reshape::melt(GenomicRanges::reduce((
          GenomicRanges::split(tes.bins, ~ gene_id)
        ),
        min.gapwidth =
          pas_window)),
        keep.extra.columns = TRUE
      )
    )
  tes.base <-
    tibble::as_tibble(tes.base)  %>%  dplyr::group_by(value.group_name) %>%
    dplyr::mutate(count =  paste0(value.group_name, ":TE", sprintf("%02d", sequence(dplyr::n(
    ))))) %>%
    GenomicRanges::makeGRangesFromDataFrame(., keep.extra.columns = TRUE)
  # assign tes_ids to isoforms
  ii <- findOverlaps(tes.bins, tes.base, maxgap = pas_window - 1)
  tes.bins.annot <-
    GenomicRanges::makeGRangesFromDataFrame(rbind(data.frame(tes.bins[S4Vectors::queryHits(ii)],
                                                             tes.base[S4Vectors::subjectHits(ii)])), keep.extra.columns = TRUE)
  tes.bins.annot <-
    tes.bins.annot %>% dplyr::mutate(transcript_id = transcript_id, tes_id =count) %>%
    plyranges::select(gene_id, transcript_id, tes_id)
  bins$tes.bins <- tes.bins
  bins$tes.base <- tes.base
  # create link database
  linksDbs <- dplyr::left_join(as.data.frame(tes.bins.annot),
                               as.data.frame(tss.bins.annot),
                               by = "transcript_id") %>%
    dplyr::select(gene_id.x, transcript_id, promoter_id, tes_id) %>% dplyr::rename(gene_id = gene_id.x) %>%
    dplyr::mutate(pairs_id = paste(promoter_id, gsub(".*:", "", tes_id), sep =":"))
  # Classify promoter type and utr type
  # sorted with proximal first distal later
  # output: class all 3'end isoforms
  le.sort <- strandSort(le)
  tt <- linksDbs %>% dplyr::group_by(gene_id) %>%
    dplyr::mutate(
      utr_type = dplyr::case_when(
        tes_id == min(tes_id) ~ "proximal",
        tes_id == max(tes_id) ~ "distal",
        TRUE ~ "other"
      )
    )
  tt <- tt %>% group_by(gene_id) %>%
    dplyr::mutate(
      promoter_type = case_when(
        promoter_id == min(promoter_id) ~ "distal",
        promoter_id == max(promoter_id) ~ "proximal",
        TRUE ~ "intermediate"
      )
    )
  # classify APA-ATSS genes
  # genes with more than 1 promoter different promoter
  atss.gene <- tt %>% dplyr::distinct(promoter_id, .keep_all = TRUE) %>%
    dplyr::group_by(gene_id) %>%  dplyr::filter(dplyr::n() > 1)  %>% dplyr::pull(gene_id)
  apa.gene <- tt %>% dplyr::distinct(tes_id, .keep_all = TRUE) %>%
    dplyr::group_by(gene_id) %>%  dplyr::filter(dplyr::n() > 1)  %>% dplyr::pull(gene_id)
  tt <-
    tt %>% dplyr::mutate(
      tss.status = ifelse(gene_id %in% atss.gene, "ATSS", "single_promoter"),
      apa.status = ifelse(gene_id %in% apa.gene, "APA", "noAPA")
    )
  # # # remove neibouring genes missassignments
  tt$tes_gene <- gsub("\\:.*", "", tt$tes_id)
  tt <- subset(tt, gene_id == tes_gene)
  tt$tes_gene <- NULL
  tt$promoter_gene <- gsub("\\:.*", "", tt$promoter_id)
  tt <- subset(tt, gene_id == promoter_gene)
  tt$promoter_gene <- NULL
  bins$links <- tt
  tt <- tt %>% dplyr::group_by(gene_id) %>%
    dplyr::mutate(
      utr_type = dplyr::case_when(
        tes_id == min(tes_id) ~ "proximal",
        tes_id == max(tes_id) ~ "distal",
        TRUE ~ "other"
      )
    )
  tt <- tt %>% group_by(gene_id) %>%
    dplyr::mutate(
      promoter_type = case_when(
        promoter_id == min(promoter_id) ~ "distal",
        promoter_id == max(promoter_id) ~ "proximal",
        TRUE ~ "intermediate"
      )
    )
  atss.gene <-  tt %>% dplyr::distinct(promoter_id, .keep_all = TRUE) %>%
    dplyr::group_by(gene_id) %>%  dplyr::filter(dplyr::n() > 1)  %>% dplyr::pull(gene_id)
  apa.gene <- tt %>% dplyr::distinct(tes_id, .keep_all = TRUE) %>%
    dplyr::group_by(gene_id) %>%  dplyr::filter(dplyr::n() > 1)  %>% dplyr::pull(gene_id)
  tt <-
    tt %>% dplyr::mutate(
      tss.status = ifelse(gene_id %in% atss.gene, "ATSS", "single_promoter"),
      apa.status = ifelse(gene_id %in% apa.gene, "APA", "noAPA")
    )
  # add pairs coordiantes
  # retrieve pairs coordinates
  # add ids to pairs
  tes.bins2 <- tes.bins %>%
    as.data.frame(.) %>%
    left_join(., tt , by= "transcript_id") %>%
    makeGRangesFromDataFrame(., keep.extra.columns = TRUE)
  tss.bins2 <- tss.bins %>%
    as.data.frame(.) %>%
    left_join(., tt , by= "transcript_id") %>%
    makeGRangesFromDataFrame(., keep.extra.columns = TRUE)
  bins2pairs <- c(tes.bins2, tss.bins2)
  pairs_range <- unlist(range(split(bins2pairs, f=bins2pairs$pairs_id)))
  pairs_range$pairs_id <- names(pairs_range)
  pairs_range$gene_id <- gsub("\\:.*","",names(pairs_range))
  result <- list()
  result$pairDataBase <- tt
  result$TESCoordinate.bins <- tes.bins
  result$TESCoordinate.base <- tes.base
  result$TSSCoordinate.bins <- tss.bins
  result$TSSCoordinate.base <- tss.base
  result$TSS_TES_isoforms.regions <- pairs_range
  return(result)
}

```


## Gene Classification based on 5'-3' regulation 

`getIsoformFeatures` allows classification of genes by they're regulatory potentitial. Similar function is used in LASER and LATER for isoform quantification and analysis.  

```{r, message=FALSE}
isoform_database <- getIsoformFeatures(cia_annot, 
                   promoter_window = 50,
                   pas_window = 150)

```

the created `isoform_database` object contains relevant objects for exploration: 

1) pairDataBase (dataframe) : Include all isoforms of the transcriptome, as well as additional categories, `promoter_id` and `tes_id` which is the identifier of the promoter/pas_site of the gene according to position. `pairs_d` which is the combination of both. Importantly the `tss.status` and `tss.status` indicate the regulation of the gene if in overall contains more than 1 alternative promoter/PAS site. 

2) Genomic positions can be found in `TESCoordinate.base` and `TSSCoordinate.base` ( GenomicRanges Objects), where the `promoter_id` and `tes_id` are assigned to they're respective genomic coordinate. 

## Identify gene types 

Gene information can be easily retrieved. As shown in the example bellow the gene `FBgn0266521` contains 18 isoforms and its classifyed as and ATSS-APA gene. Traceable promoter and tes_id can be also found in the table. 

```{r}
isoform_database$pairDataBase %>% 
  filter(gene_id == "FBgn0266521")
```


## Retrieve regions of interest

To retrieve relevant genomic regions is possible just to index with the gene_id as shown bellow: This will retrieve all idenitfied PAsites of the gene. 

```{r}
isoform_database$TESCoordinate.base[isoform_database$TESCoordinate.base$value.group_name  == "FBgn0266521"]
```
Similar approach can be used to retrieve promoter coordinates 
```{r}
isoform_database$TSSCoordinate.base[isoform_database$TSSCoordinate.base$value.group_name  == "FBgn0266521"]

```

# Retrieve 5'-3' isoforms associated with a given gene. 

Additionally the full length 5'-3' isoform region can be retrieve for future studies. 

```{r}
isoform_database$TSS_TES_isoforms.regions[isoform_database$TSS_TES_isoforms.regions$gene_id  == "FBgn0266521"]
```



