---
title: "Start Sites detection with ONP-PCR"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(require(GenomicRanges))
suppressPackageStartupMessages(require(rtracklayer))
suppressPackageStartupMessages(require(edgeR))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(ggsci))
suppressPackageStartupMessages(require(ChIPseeker))
suppressPackageStartupMessages(require(TxDb.Dmelanogaster.UCSC.dm6.ensGene))
txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene

```

## Load data 

```{r, include = FALSE }
#path refs 
refList <- list()
# ensembl 
pathRefAnnot <- "/data/repository/organisms/dm6_ensembl/ensembl/release-96/genes.gtf"
refList$exon <- rtracklayer::import.gff(pathRefAnnot, feature.type="exon")
# promoter edp 
edpPath <- "/data/hilgers/group/alfonso/projects/LongReadSeq/repo/promoter.db.edp.bed"
refList$edpDb <- rtracklayer::import.bed(edpPath)
# make TSS ref from ensembl  
#ensDataBase <- LORD::prepareLinksDatabase(refList$exon , tss.window = 50, tes.window = 100)
dir5ends <- "/data/hilgers/group/alfonso/projects/2021_LRS_paper/data/dominant_promoters_dmel.NewCounter/bed/"
endsPath <- list.files(dir5ends, pattern = "*5p.read_ends.bed", full.names = TRUE)
bedFiles <- setNames( lapply(endsPath, rtracklayer::import.bed), n = basename(endsPath)) 
# keep cannonical chromosomes 
bedFiles <- lapply(bedFiles, function(x){ 
  x[seqnames(x) %in% c("X", "3L" , "2L" , "3R" , "2R", "4"),]})
```

# Functions
```{r}

```


## Make peaks and count 
```{r , echo=FALSE}
# make peaks from reads 
peaks <- lapply(bedFiles, function (x) IRanges::reduce(x, min.gapwidth=150)  )
# count peaks 
peakCounts <- function(x,y){
  x$peak.counts <- countOverlaps(x,y)
  x$cpms <- edgeR::cpm(x$peak.counts)
  return(x)
  }
peakCounts <- mapply(peakCounts, peaks, bedFiles) 
```

## peak counts and overlaps with Promoter database. window 100 nt 

```{r}
promoterDataBase <- refList$edpDb
countsParams <- seq(1,50)
filterCompare <- function(countCut, peakRegions){ 
  fCounts <- peakRegions[peakRegions$peak.counts>countCut,]
  hits <- findOverlaps(fCounts,promoterDataBase, maxgap = 150)
  length(queryHits(hits))/queryLength(hits)
}
lapply.filterCompare<- function(peaks, countsCut){ 
  do.call(rbind, setNames( lapply(countsCut, filterCompare, peaks), n = paste0("counts:",countsCut)))
}
peaksPerCounts <-  lapply( peakCounts, lapply.filterCompare, countsCut= countsParams)  
names(peaksPerCounts) <- c("Heads_", "embryos.pooled_", "ovaries_", "embryos2_")
fractionOfPeaks <- do.call(rbind, lapply(peaksPerCounts, function(x){ as.data.frame(x)}) ) %>% mutate(tissue=gsub("\\_.*","",rownames(.)), countCut=gsub(".*_.","",rownames(.))) 

fractionOfPeaks %>% mutate(counts=as.numeric(gsub("counts:", "", countCut)) ) %>% filter(!tissue=="embryos2") %>% ggplot(., aes(x=counts, y=V1, colour=tissue)) + geom_point() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Fraction TSS overalps w EDP") + theme_classic() + xlab("min Read counts") + scale_color_aaas()

```

## peak counts and overlaps with Promoter database. window 100 nt 

```{r}
norm.countsParams <- seq(0,50)
filterCompare <- function(countCut, peakRegions){ 
  fCounts <- peakRegions[peakRegions$cpms>countCut,]
  hits <- findOverlaps(fCounts,promoterDataBase, maxgap = 150)
  length(queryHits(hits))/queryLength(hits)
}
lapply.filterCompare<- function(peaks, countsCut){ 
  do.call(rbind, setNames( lapply(countsCut, filterCompare, peaks), n = paste0("counts:",countsCut)))
}
peaksPerCounts <-  lapply( peakCounts, lapply.filterCompare, countsCut= countsParams)  
names(peaksPerCounts) <- c("Heads_", "embryos.pooled_", "ovaries_", "embryos2_")
fractionOfPeaks <- do.call(rbind, lapply(peaksPerCounts, function(x){ as.data.frame(x)}) ) %>% mutate(tissue=gsub("\\_.*","",rownames(.)), countCut=gsub(".*_.","",rownames(.)))
```


```{r}
fractionOfPeaks %>% mutate(counts=as.numeric(gsub("counts:", "", countCut)) ) %>% filter(!tissue=="embryos2") %>% ggplot(., aes(x=counts, y=V1, colour=tissue)) + geom_point() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Fraction TSS overalps w EDP") + theme_classic() + xlab("min counts per million") + scale_color_aaas() + geom_vline(xintercept = 15) 

```

# compared to ensembl 
```{r}
#ensDataBase <- LORD::prepareLinksDatabase(refList$exon , tss.window = 50, tes.window = 100)
```

# 

# peaks bellow expression threshould
```{r, include=FALSE}
subsetByOverlapEDP <- function(peakRegions){ 
  fCounts <- peakRegions[peakRegions$cpms<15,]
  hits <- findOverlaps(fCounts,promoterDataBase, maxgap = 150)
  hitEDP<- fCounts[-queryHits(hits),]
  return(hitEDP)
}
nonOverlappingPeaksLow <- lapply(peakCounts, subsetByOverlapEDP)

```

```{r, message=FALSE, warning=FALSE}
nonOverlappingPeaksLow.ucsc <- lapply(nonOverlappingPeaksLow, function(x){
  seqlevelsStyle(x) <- "UCSC"
  return(x)
})
annotPeaksPerTissue<- lapply(nonOverlappingPeaksLow.ucsc, function(x){
   annotatePeak(x, tssRegion=c(-150, 150),
                         TxDb=txdb, annoDb="org.Dm.eg.db")
})
```

# heads 
```{r}
plotAnnoPie(annotPeaksPerTissue$onp_1D_heads_pooled.bed5p.read_ends.bed, )
```

# embryos 
```{r}
plotAnnoPie(annotPeaksPerTissue$w1118_embryos_pooled.bed5p.read_ends.bed)
```

#ovaries
```{r}
plotAnnoPie(annotPeaksPerTissue$w118_ovaries_virgin_females_pooled.bed5p.read_ends.bed)

```


# After filtering of 15 CPMs 

```{r}
filterSummarize <- function(peakRegions){ 
  fCounts <- peakRegions[peakRegions$cpms>15,]
  hits <- findOverlaps(fCounts,promoterDataBase, maxgap = 100)
  peakOverlapTable <- data.frame(peakTypes=c("total_peaks", "peaksInRef", "peaksOnlyLR"), n=c(queryLength(hits),length(queryHits(hits)),queryLength(hits)-length(queryHits(hits)) ) )
  return(peakOverlapTable)
}

do.call(rbind, setNames( lapply(peakCounts, filterSummarize),n=c("Heads", "Embryos", "Ovaries", "Embryo2")) ) %>% mutate(tissue=gsub("\\..*","",rownames(.)) ) %>% filter(!tissue=="Embryo2") %>% ggplot(data=., aes(x=peakTypes, y=n, fill=peakTypes)) +
  geom_bar(stat="identity") + facet_grid(cols=vars(tissue)) + theme_classic() + scale_fill_aaas() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + ylab("Number of peaks")

```


## annotate non overlapping peaks 

```{r}
subsetByOverlapEDP <- function(peakRegions){ 
  fCounts <- peakRegions[peakRegions$cpms>15,]
  hits <- findOverlaps(fCounts,promoterDataBase, maxgap = 150)
  hitEDP<- fCounts[-queryHits(hits),]
  return(hitEDP)
}
nonOverlappingPeaksHigh <- lapply(peakCounts, subsetByOverlapEDP)
```

```{r, include=FALSE, warning=FALSE, message=FALSE}
nonOverlappingPeaksHigh.ucsc <- lapply(nonOverlappingPeaksHigh, function(x){
  seqlevelsStyle(x) <- "UCSC"
  return(x)
})
annotPeaksPerTissue<- lapply(nonOverlappingPeaksHigh.ucsc, function(x){
   annotatePeak(x, tssRegion=c(-150, 150),
                         TxDb=txdb, annoDb="org.Dm.eg.db")
})

```

# per tissue regions 
# head 
```{r}
plotAnnoPie(annotPeaksPerTissue$onp_1D_heads_pooled.bed5p.read_ends.bed)
```

# embryo 
```{r}
plotAnnoPie(annotPeaksPerTissue$w1118_embryos_pooled.bed5p.read_ends.bed)
```

# ovaries 
```{r}
plotAnnoPie(annotPeaksPerTissue$w118_ovaries_virgin_females_pooled.bed5p.read_ends.bed)
```




